src/turbomind/models/llama/LlamaV2.cc

```diff
@@ -252,18 +252,27 @@ void LlamaV2<T>::forwardUnified(T*               out,
 {
     TM_LOG_DEBUG(__PRETTY_FUNCTION__);
 
+    int table_length = vocab_size_padded_ / tensor_para_.world_size_;
+    int table_offset = table_length * tensor_para_.rank_;
     invokeInputIdsEmbeddingLookupPosEncoding(decoder_input,
                                              nullptr,  // processed somewhere else
                                              weights_->pre_decoder_embedding_table,
                                              static_cast<T*>(nullptr),
                                              pPromptTuningParam<T>{},
                                              input_ids,
+                                             table_offset,
+                                             table_length,
                                              0,  // only used for position encoding
                                              token_num,
                                              token_num,
                                              1,
                                              hidden_units_,
                                              stream_);
+    if (tensor_para_.world_size_ > 1) {
+        NcclGuard nccl_guard(tensor_para_, stream_);
+        ftNcclAllReduceSum(decoder_input, decoder_input, token_num * hidden_units_, tensor_para_, stream_);
```

Comment:
Why use allreduce? Shouldn't we use allgather?

Reply:
If using allgather, the weight and buffer layout should change and after allgather it need another permute op.

And another reason which confuses me is in my test, the reduce sum is faster than allgather.

Reply:
@lzhangzz what's the permute op in turbomind
In my opinion, it is more reasonable to use 'allgather' than to use 'allreduce'.
@lzhangzz do you have any idea about "the reduce sum is faster than allgather"